perf/src/qos.rs

```diff
@@ -0,0 +1,121 @@
+use {
+    crate::{
+        data_budget::DataBudget,
+        packet::{Packet, PacketBatch},
+    },
+    std::{
+        collections::HashMap,
+        net::IpAddr,
+        sync::atomic::{AtomicUsize, Ordering},
+    },
+};
+
+pub type StakeMap = HashMap<IpAddr, (u64, DataBudget)>;
+
+fn qos_packet(
+    num_discarded: &AtomicUsize,
+    max_packets_per_second: u64,
+    default_budget: &DataBudget,
+    total_stake: u64,
+    ip_to_budget: &StakeMap,
+    packet: &mut Packet,
+) {
+    if packet.meta.discard() {
+        return;
+    }
+
+    if let Some((stake, budget)) = ip_to_budget.get(&packet.meta.addr) {
+        budget.update(1000, |_| {
+            (*stake as u128 * max_packets_per_second as u128 / total_stake as u128) as usize
+        });
+        if !budget.take(1) {
+            packet.meta.set_discard(true);
+            num_discarded.fetch_add(1, Ordering::Relaxed);
+        }
+    } else {
+        default_budget.update(1000, |_| max_packets_per_second as usize / 10);
```

Comment:
how much is given to unstaked traffic?  If its equal to this nodes stake weight, it would be balanced across the network.

Reply:
10% in this